File: app/src/main/java/com/salman/bitclock/receivers/BootCompletedReceiver.java
package com.salman.bitclock.receivers;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;

import com.salman.bitclock.data.repository.AlarmRepository;
import com.salman.bitclock.data.models.Alarm;
import com.salman.bitclock.di.AppModule;
import com.salman.bitclock.utils.AlarmScheduler;

import java.util.List;

import dagger.hilt.android.EntryPointAccessors;

/**
 * Re-schedules alarms when the device boots up.
 * Optimized to use synchronous database calls in a background thread.
 */
public class BootCompletedReceiver extends BroadcastReceiver {

    private static final String TAG = "BootCompletedReceiver";

    @Override
    public void onReceive(Context context, Intent intent) {
        if (Intent.ACTION_BOOT_COMPLETED.equals(intent.getAction())) {
            
            final PendingResult pendingResult = goAsync();
            
            AppModule.AlarmRepositoryEntryPoint entryPoint = EntryPointAccessors.fromApplication(
                    context, AppModule.AlarmRepositoryEntryPoint.class);
            AlarmRepository alarmRepository = entryPoint.alarmRepository();
            AlarmScheduler scheduler = new AlarmScheduler(context);

            new Thread(() -> {
                try {
                    // Fetch alarms synchronously from the database
                    List<Alarm> alarms = alarmRepository.getAllAlarmsSync();
                    for (Alarm alarm : alarms) {
                        if (alarm.isEnabled()) {
                            scheduler.scheduleAlarm(alarm);
                        }
                    }
                } finally {
                    pendingResult.finish();
                }
            }).start();
        }
    }
}

File: app/src/main/kotlin/com/salman/bitclock/receivers/BootCompletedReceiver.kt
package com.salman.bitclock.receivers

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.util.Log
import com.salman.bitclock.di.AppModule
import com.salman.bitclock.utils.AlarmScheduler
import dagger.hilt.android.EntryPointAccessors
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

class BootCompletedReceiver : BroadcastReceiver() {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val TAG = "BootCompletedReceiver"

    override fun onReceive(context: Context, intent: Intent?) {
        val action = intent?.action
        Log.d(TAG, "Received broadcast: $action")
        
        val bootActions = listOf(
            Intent.ACTION_BOOT_COMPLETED,
            "android.intent.action.QUICKBOOT_POWERON",
            "com.htc.intent.action.QUICKBOOT_POWERON"
        )

        if (action in bootActions) {
            val pendingResult = goAsync()
            val entryPoint = EntryPointAccessors.fromApplication(
                context.applicationContext, AppModule.AlarmRepositoryEntryPoint::class.java
            )
            val repository = entryPoint.alarmRepository()
            val scheduler = AlarmScheduler(context)

            scope.launch {
                try {
                    val alarms = repository.getAllAlarmsSync()
                    Log.d(TAG, "Rescheduling ${alarms.count { it.isEnabled }} enabled alarms")
                    for (alarm in alarms) {
                        if (alarm.isEnabled) {
                            scheduler.scheduleAlarm(alarm)
                        }
                    }
                } catch (e: Exception) {
                    Log.e(TAG, "Error rescheduling alarms after boot", e)
                } finally {
                    pendingResult.finish()
                }
            }
        }
    }
}
